File: ZhuZhiLiao/Motion/MotionSampleFilter.swift
import simd

struct MotionFilterConfiguration: Equatable, Sendable {
    var deadZone: Float
    var maximumMagnitude: Float
    var smoothingFactor: Float

    static let live = MotionFilterConfiguration(
        deadZone: 0.03,
        maximumMagnitude: 2.5,
        smoothingFactor: 0.25
    )
}

struct MotionSampleFilter: Sendable {
    let configuration: MotionFilterConfiguration
    private(set) var filteredValue = SIMD3<Float>.zero

    init(configuration: MotionFilterConfiguration = .live) {
        self.configuration = configuration
    }

    mutating func process(_ sample: SIMD3<Float>) -> SIMD3<Float> {
        let magnitude = simd_length(sample)
        let target: SIMD3<Float>

        if magnitude < configuration.deadZone {
            target = .zero
        } else if magnitude > configuration.maximumMagnitude, magnitude > 0 {
            target = sample * (configuration.maximumMagnitude / magnitude)
        } else {
            target = sample
        }

        let smoothing = min(max(configuration.smoothingFactor, 0), 1)
        filteredValue += (target - filteredValue) * smoothing
        return filteredValue
    }

    mutating func reset() {
        filteredValue = .zero
    }
}

File: Android/app/src/main/java/com/azhegezhege/zhuzhiliao/motion/MotionSampleFilter.kt
package com.azhegezhege.zhuzhiliao.motion

import com.azhegezhege.zhuzhiliao.math.Vec3

data class MotionFilterConfiguration(
    val deadZone: Float = 0.03f,
    val maximumMagnitude: Float = 2.5f,
    val smoothingFactor: Float = 0.25f,
)

class MotionSampleFilter(
    val configuration: MotionFilterConfiguration = MotionFilterConfiguration(),
) {
    var filteredValue: Vec3 = Vec3.ZERO
        private set
    private var lastTimestamp: Double? = null

    fun process(sample: Vec3, timestamp: Double? = null): Vec3 {
        if (timestamp != null && timestamp == lastTimestamp) return filteredValue
        val magnitude = sample.length
        val target = when {
            magnitude < configuration.deadZone -> Vec3.ZERO
            magnitude > configuration.maximumMagnitude && magnitude > 0f ->
                sample * (configuration.maximumMagnitude / magnitude)
            else -> sample
        }
        val smoothing = configuration.smoothingFactor.coerceIn(0f, 1f)
        filteredValue += (target - filteredValue) * smoothing
        lastTimestamp = timestamp
        return filteredValue
    }

    fun reset() {
        filteredValue = Vec3.ZERO
        lastTimestamp = null
    }
}
